src/repl.rs

```diff
@@ -76,25 +82,55 @@ fn find_info(args: WithVM<RootStr>) -> IO<Result<String, String>> {
     IO::Value(Ok(buffer))
 }
 
-fn compile_repl(vm: &Thread) -> Result<(), Box<StdError + Send + Sync>> {
-    fn input(prompt: &str) -> IO<Option<String>> {
-        let input = ::linenoise::input(prompt);
+struct Editor(Mutex<::rustyline::Editor<()>>);
 
-        if let Some(ref input) = input {
-            if !input.trim().is_empty() {
-                ::linenoise::history_add(input);
-            }
-        }
+impl fmt::Debug for Editor {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "File")
+    }
+}
+
+impl VMType for Editor {
+    type Type = Editor;
+}
+
+impl Traverseable for Editor {
+    fn traverse(&self, _: &mut Gc) {}
+}
+
+fn new_editor(_: ()) -> Userdata<Editor> {
+    let editor = ::rustyline::Editor::new();
+    Userdata(Editor(Mutex::new(editor)))
+}
 
-        IO::Value(input)
+fn readline(editor: &Editor, prompt: &str) -> IO<Option<String>> {
+    let mut editor = editor.0.lock().unwrap();
+    let input = match editor.readline(prompt) {
+        Ok(input) => input,
+        Err(ReadlineError::Eof) => return IO::Value(None),
+        Err(err) => return IO::Exception(format!("{}", err)),
```

Comment:
Ending the program on `ReadlineError::Eof` (ie. ^D) makes sense - but what about `ReadlineError::Interrupted` (ie. ^C)? I would say if it is easiest, return `IO::Value(None)`, but in the future it would be nice for it to kill a long-running expression.

https://kkawakam.github.io/rustyline/rustyline/error/enum.ReadlineError.html